crate/findex/src/batcher_findex.rs

```diff
@@ -0,0 +1,299 @@
+use std::{collections::HashSet, fmt::Debug, hash::Hash, sync::Arc};
+
+use cosmian_sse_memories::{ADDRESS_LENGTH, Address, BatchingMemoryADT, MemoryBatcher};
+
+use crate::{Decoder, Encoder, Findex, IndexADT, adt::IndexBatcher, error::BatchFindexError};
+
+#[derive(Debug)]
+pub struct FindexBatcher<
+    const WORD_LENGTH: usize,
+    Value: Send + Hash + Eq,
+    EncodingError: Send + Debug,
+    BatcherMemory: Clone + Send + BatchingMemoryADT<Address = Address<ADDRESS_LENGTH>, Word = [u8; WORD_LENGTH]>,
+> {
+    memory: BatcherMemory,
+    encode: Encoder<Value, BatcherMemory::Word, EncodingError>,
+    decode: Decoder<Value, BatcherMemory::Word, EncodingError>,
+}
+
+impl<
+    const WORD_LENGTH: usize,
+    Value: Send + Hash + Eq,
+    BatcherMemory: Debug
+        + Send
+        + Sync
+        + Clone
+        + BatchingMemoryADT<Address = Address<ADDRESS_LENGTH>, Word = [u8; WORD_LENGTH]>,
+    EncodingError: Send + Debug,
+> FindexBatcher<WORD_LENGTH, Value, EncodingError, BatcherMemory>
+{
+    pub fn new(
+        memory: BatcherMemory,
+        encode: Encoder<Value, BatcherMemory::Word, EncodingError>,
+        decode: Decoder<Value, BatcherMemory::Word, EncodingError>,
+    ) -> Self {
+        Self {
+            memory,
+            encode,
+            decode,
+        }
+    }
+
+    // Both insert and delete operations make an unbounded number of calls to
+    // `guarded_write` on the memory layer.
+    async fn batch_insert_or_delete<Keyword>(
+        &self,
+        entries: Vec<(Keyword, impl Send + IntoIterator<Item = Value>)>,
+        is_insert: bool,
+    ) -> Result<(), BatchFindexError<BatcherMemory>>
+    where
+        Keyword: Send + Sync + Hash + Eq,
+    {
+        let mut futures = Vec::new();
+        let memory = Arc::new(MemoryBatcher::new(self.memory.clone(), entries.len()));
+
+        for (guard_keyword, bindings) in entries {
+            let memory = memory.clone();
+            // Create a temporary Findex instance using the shared batching layer.
+            let findex = Findex::<WORD_LENGTH, Value, EncodingError, _>::new(
+                memory.clone(),
+                self.encode,
+                self.decode,
+            );
+
+            let future = async move {
+                if is_insert {
+                    findex.insert(guard_keyword, bindings).await
+                } else {
+                    findex.delete(guard_keyword, bindings).await
+                }?;
+                // Once one of the operations succeeds, we should make the buffer smaller.
```

Comment:
The concept of buffer is an internal of the memory batcher, which makes this comment hard to understand without knowing the memory batcher implementation.
```suggestion
                // Once an operation succeeds, the batch size needs to be reduced.
```

Reply:
done